### PR TITLE
Add exclude_packages_in_use option to JamfPackageCleaner

### PR DESCRIPTION
**Pull request by smithjw**
JamfPackageCleaner selects packages to delete purely by name match and package ID order, so an older package that is still referenced by a policy, patch software title, or PreStage Enrollment is deleted once enough newer versions exist. That breaks any policy pinned to the older package.

Add an opt-in `exclude_packages_in_use` input (default `False`, so existing behaviour is unchanged). When `True`, any package still in use is kept even if it would otherwise be deleted. The `versions_to_keep` newest packages are always kept, so the option only ever spares additional older packages. The usage lookup is skipped when there is nothing to delete, to avoid enumerating every policy, patch title, and PreStage Enrollment unnecessarily. The token is refreshed before each lookup (as the deletion loop does) so it cannot expire mid-enumeration on large servers.

Move `get_packages_in_policies`, `get_packages_in_patch_titles`, and `get_packages_in_prestages` from `JamfUnusedPackageCleanerBase` up into `JamfUploaderBase` so both cleaners share one implementation.

Add a stubbed test exercising the filter, the unchanged default, and the performance guard.

### CODE REVIEW

File: JamfUploaderProcessors/JamfUploaderLib/JamfUploaderBase.py [hunk 1]
```diff
@@ -2294,6 +2294,125 @@ def get_api_object_value_from_id(
...
+                    tenant_id=tenant_id,
+                )
+                try:
+                    pkgs = generic_info["package_configuration"]["packages"]
+                    for x in pkgs:
+                        pkg = x["name"]
+                        if pkg not in packages_in_policies:
+                            packages_in_policies.append(pkg)
+                except IndexError:
+                    pass
```
[the remark this answers is not in the document]
Comment (smithjw) [reply JamfUploaderProcessors/JamfUploaderLib/JamfUploaderBase.py]: Fixed in 8886fcd. The catch now includes KeyError and TypeError, so a policy with no `package_configuration` counts as having no packages instead of crashing.

File: JamfUploaderProcessors/JamfUploaderLib/JamfUploaderBase.py [hunk 2]
```diff
@@ -2294,6 +2294,125 @@ def get_api_object_value_from_id(
...
+                                pkg = versions[i]["package"]["name"]
+                                if pkg:
+                                    if pkg != "None" and pkg not in packages_in_titles:
+                                        packages_in_titles.append(pkg)
+                            except IndexError:
+                                pass
+                            except KeyError:
+                                pass
+                except IndexError:
+                    pass
```
[the remark this answers is not in the document]
Comment (smithjw) [reply JamfUploaderProcessors/JamfUploaderLib/JamfUploaderBase.py:2398]: Fixed in 8886fcd by catching `TypeError` so versions coming back as `None` no longer aborts the run

File: JamfUploaderProcessors/JamfUploaderLib/JamfUploaderBase.py [hunk 3]
```diff
@@ -2294,6 +2294,125 @@ def get_api_object_value_from_id(
...
+                (
+                    "Please wait while we gather a list of all packages in all "
+                    f"PreStage Enrollments (total {len(prestages)})..."
+                ),
+                verbose_level=1,
+            )
+            for _, prestage in enumerate(prestages):
+                pkg_ids = prestage["customPackageIds"]
+                if len(pkg_ids) > 0:
+                    for pkg_id in pkg_ids:
```
[the remark this answers is not in the document]
Comment (smithjw) [reply JamfUploaderProcessors/JamfUploaderLib/JamfUploaderBase.py:2424]: Also fixed in 8886fcd. Used `.get("customPackageIds")` or `[]` as suggested so a missing key or `null` value just means no packages

**Review by grahampugh — COMMENTED** [review]
## PR #385 — Add `exclude_packages_in_use` option to `JamfPackageCleaner`

**Scope of review:** Full diff across 7 files (~350 lines changed + 212-line test). Copilot flagged robustness issues in the pre-fix iteration; all three were addressed in commit `8886fcd`.

---

### What this PR does

Adds an opt-in `exclude_packages_in_use` flag to `JamfPackageCleaner` so packages still referenced by policies, patch titles, or PreStage Enrollments are spared from deletion. As a housekeeping step it promotes the three `get_packages_in_*` helpers from `JamfUnusedPackageCleanerBase` up into the shared `JamfUploaderBase` so both cleaners can use them.

---

### Findings

❌ **CRITICAL (1) (correctness) — Keyword argument mismatch breaks `JamfUnusedPackageCleaner`**

The refactor moves `get_packages_in_policies`, `get_packages_in_patch_titles`, and `get_packages_in_prestages` from `JamfUnusedPackageCleanerBase` into `JamfUploaderBase` with the keyword parameter renamed from `platform_level_id` to `tenant_id`. But `JamfUnusedPackageCleanerBase.execute()` — which is **not** updated in this PR — still calls all three methods using the old keyword:

```python
# JamfUnusedPackageCleanerBase.execute() — unchanged by this PR (lines ~450-459 on main)
packages_in_prestages = self.get_packages_in_prestages(
    api_url, token, platform_level_id=platform_level_id   # ← old keyword
)
packages_in_titles = self.get_packages_in_patch_titles(
    api_url, token, platform_level_id=platform_level_id
)
packages_in_policies = self.get_packages_in_policies(
    api_url, token, platform_level_id=platform_level_id
)
```

After this PR the method resolution lands on the new `JamfUploaderBase` implementations (signature `def get_packages_in_policies(self, api_url, token, tenant_id="")`), which don't accept `platform_level_id=`. Python will raise:

```
TypeError: get_packages_in_policies() got an unexpected keyword argument 'platform_level_id'
```

This means `JamfUnusedPackageCleaner` is completely broken post-merge on any instance that has packages to enumerate.

**Fix:** Update the three call sites in `JamfUnusedPackageCleanerBase.execute()` to pass `tenant_id=` instead of `platform_level_id=` — matching what `JamfPackageCleanerBase` already does in the new code.

**Acceptance criteria:** A dry run of `JamfUnusedPackageCleaner` against an instance with existing packages must not raise `TypeError`.

---

⚠️ **IMPORTANT (2) (test coverage) — Test doesn't catch the breaking regression**

`StubCleaner` overrides all three `get_packages_in_*` methods, so the test exercises `JamfPackageCleaner`'s selection logic in isolation but provides no coverage of `JamfUnusedPackageCleaner` calling through the shared implementations. Had even a smoke-level test for `JamfUnusedPackageCleaner` existed, finding #1 would have surfaced immediately.

**Fix:** Add a `StubUnusedCleaner` analogous to `StubCleaner` and assert that it can complete an `execute()` run without raising. The existing `UsageGetterHarness` already covers the `get_packages_in_*` bodies — one happy-path run of the unused cleaner is enough to guard the inheritance chain.

**Acceptance criteria:** `_tests/test_package_cleaner.py` includes at least one `JamfUnusedPackageCleaner` execution path covering the `platform_level_id` → `tenant_id` call sites.

---

💡 **NICE-TO-HAVE (3) — `kept_in_use` not surfaced as a dedicated output variable**

`jamfpackagecleaner_summary_result["data"]["kept_in_use"]` is set correctly, but there's no dedicated `output_variables` entry in `JamfPackageCleaner.py`. Recipes that want to branch on whether any packages were spared can't do so without parsing the summary dict. Given that `dry_run` already has `dry_run_summary_result` as its own output variable, a `packages_kept_in_use` output variable would be consistent.

---

✅ **POSITIVE — Robustness improvements to the promoted helpers are genuinely better**

The old implementations used bare `except IndexError` which missed `KeyError` and `TypeError`. The new versions catch all three, use `.get("customPackageIds") or []` for the prestage case, and skip individual malformed entries rather than dropping all well-formed entries in the same policy. The `UsageGetterHarness` tests verify each edge case against the real method bodies — exactly the right level of coverage.

---

<details><summary>Review coverage</summary>

- [x] Design / architecture — method promotion is the right call; regression introduced
- [x] Correctness — critical keyword-argument break found
- [x] Security — no concerns
- [na] Performance — opt-in flag; skip-when-nothing-to-delete guard is in place
- [x] Test coverage — StubCleaner gap identified
- [na] Reliability — token refresh pattern consistent with existing deletion loop

</details>

## Rating: 2/5

The feature logic is sound and the robustness improvements are well done, but the refactor silently breaks `JamfUnusedPackageCleaner` due to the `platform_level_id` → `tenant_id` rename not being propagated to `JamfUnusedPackageCleanerBase.execute()`. One-liner fix, but a hard regression for existing users of the unused-package cleaner.



_Generated by `pr-review` v1.31.0, a [Jamf Claude Code skill](https://github.com/jamf/claude-plugins/blob/main/plugins/pr-review/README.md)_

**Comment by smithjw** [comment]
Ok there shouldn't be any more conflicts here and I've addressed the feedback... hopefully 😅

**Review by grahampugh — COMMENTED** [review]
## Re-review: PR #385 — Add `exclude_packages_in_use` option to `JamfPackageCleaner`

> [!TIP]
> All three findings from the prior review are resolved. This PR is clean and merge-ready.

## Rating: 5/5
- Clean — the NICE-TO-HAVE `packages_kept_in_use` output variable was added, and the CRITICAL/IMPORTANT fixes are solid. No open findings.
- Coverage: specialist lanes (`security-reviewer`, `performance-reviewer`, `devil-advocate`, `test-quality-reviewer`) have never run on this PR — the rating reflects the dimensions that were searched, not the whole diff.

---

## Prior findings status

| # | Location | State | Notes |
|---|----------|-------|-------|
| (1) | `JamfUploaderBase.py` `get_packages_in_*` | ✅ Fixed | Parameter renamed back to `platform_level_id=""` in all three method signatures — callers in `JamfUnusedPackageCleanerBase` now match |
| (2) | `_tests/test_package_cleaner.py` | ✅ Fixed | `StubUnusedCleaner` added; asserts correct forwarding of `platform_level_id` through the inheritance chain |
| (3) | `JamfPackageCleaner.py` `output_variables` | ✅ Fixed | `packages_kept_in_use` declared in `output_variables` and correctly set in `JamfPackageCleanerBase.execute()` |

---

<details>
<summary>Scope of review</summary>

- Incremental diff reviewed: commits since `23c3f1f6` (prior review baseline) → `f54c5e1` (current head), including an upstream merge commit.
- Most of the incremental diff is the upstream `main` merge — `CLAUDE.md` updates, `JSS_URL: required: False` changes, `PLATFORM_API_ENVIRONMENT_ID` additions, etc. These are out of scope for this PR's review.
- PR-specific fix commits addressed: parameter name reconciliation, test harness for `JamfUnusedPackageCleaner`, and `packages_kept_in_use` output variable.
- **Never-run specialist lanes** (eligible on this PR's diff, not dispatched because the PR is merge-ready and all findings are resolved):
  - `security-reviewer` — inapplicable (no auth/security surfaces changed by this PR specifically)
  - `performance-reviewer` — eligible (package enumeration loops); never ran
  - `devil-advocate` — eligible; never ran
  - `test-quality-reviewer` — eligible; never ran

</details>

---

_Review by [Jamf Claude Code skill](https://github.com/grahampugh/jamf-upload) v1.31.0_
